src/ui/PackageDependencyProvider.ts

```diff
@@ -148,38 +155,138 @@ export class PackageDependenciesProvider implements vscode.TreeDataProvider<Tree
-            // be up to date with edited dependency list, we need to remove the edited
-            // dependencies from the list before adding in the edit version
-            const children = [
-                ...this.getLocalDependencies(workspaceState),
-                ...this.getRemoteDependencies(folderContext),
-            ];
-            const editedChildren = await this.getEditedDependencies(workspaceState);
-            const uneditedChildren: PackageNode[] = [];
-            for (const child of children) {
-                const editedVersion = editedChildren.find(item => item.name === child.name);
-                if (!editedVersion) {
-                    uneditedChildren.push(child);
-                }
+            return await this.getDependencyGraph(workspaceState, folderContext);
+        }
+
+        return this.getNodesInDirectory(element.path);
+    }
+
+    private async getDependencyGraph(
+        workspaceState: WorkspaceState | undefined,
+        folderContext: FolderContext
+    ): Promise<PackageNode[]> {
+        if (!workspaceState) {
+            return [];
+        }
+        const inUseDependencies = await this.getInUseDependencies(workspaceState, folderContext);
+        return (
+            workspaceState?.object.dependencies
+                .filter(dependency => inUseDependencies.has(dependency.packageRef.identity))
+                .map(dependency => {
+                    const type = this.dependencyType(dependency);
+                    const version = this.dependencyDisplayVersion(dependency);
+                    const packagePath = this.dependencyPackagePath(
+                        dependency,
+                        folderContext.folder.fsPath
+                    );
+                    const location = dependency.packageRef.location;
+                    return new PackageNode(
+                        dependency.packageRef.identity,
+                        packagePath,
+                        location,
+                        version,
+                        type
+                    );
+                }) ?? []
+        );
+    }
+
+    /**
+     * * Returns a set of all dependencies that are in use in the workspace.
+     * Why tranverse is necessary here?
+     *  * If we have an implicit local dependency of a dependency, you may not be able to see it in either `Package.swift` or `Package.resolved` unless tranversing from root Package.swift.
+     * Why not using `swift package show-dependencies`?
+     *  * it costs more time and it triggers the file change of `Workspace-state.json` which is not necessary
+     * Why not using `workspace-state.json` directly?
+     *  * `workspace-state.json` contains all necessary dependencies but it also contains dependencies that are not in use.
+     * Here is the implementation details:
+     * 1. local/remote/edited dependency has remote/edited dependencies, Package.resolved covers them
```

Comment:
@adam-fowler 

After I tried more cases, I just found item 1 is wrong. originally I saw edited remote package is  in `Package.resolved` <-- but it's not . perhaps I prepared a wrong test case
the test case is 
1. A depends on B, B depends on C
2. `swift package edit C`, **C will dissapear from `Package.resolved`**

This means we can only check if C is in use by traversing from root or we just show `edited` dependencies out anyway. 

then we have to go back to the original discuss point. 
It seems traversing is the only way to get correct dependencies graph

Currently I added `getEditedDependencySet` which read edited packages from `workspace-state.json`, **but it's still wrong**. because once A is removed from `Package.swift`.  C is still in `workspace-state.json` as `edited`. so it will still be in the list  (deleting state.json could fix the issue)

lemme summarize a bit again 
1. tranverse from root, this is the best result [ no missing items, no extra items, but concern about performance]
2. load dependencies from workspace-state.json [no missing items, but sometimes extra items, no performance issue (deleting state.json could fix the issue)]
3. current PR, traversing local dependencies + reading editing from state.json + remote packages from pin [extra editing packages sometimes (deleting state.json could fix the issue)]

my preference is 1 >= 2 >= 3

What do you think

Reply:
1. This is a real performance issue
2. Yeah workspace-state.json would be the best solution, but it isn't kept up to date. I don't know what the implications of deleting the workspace-state.json every time we resolve is. I have added this issue to the SwiftPM repo https://github.com/apple/swift-package-manager/issues/5727
3. I'm happy to go with this temporarily. It improves on our current situation.

Reply:
I also created a PR https://github.com/apple/swift-package-manager/pull/5700/
if it can be accepted, workspace-state.json file could be watched in future

https://github.com/apple/swift-package-manager/issues/5727 I'm not sure if they will treat it as a bug, because they may still treat the unused dependencies as "`ManagedDepdency`",  hopefully, they can fix this issue